src/index.ts

```diff
@@ -1,1692 +1,3213 @@
 
-    //Set Inverter Status Message and dot
-    let inverterStateColour = "";
-    let inverterStateMsg = "";
-    let inverterModel = "";
-    let found = false;
+        //Set Inverter Status Message and dot
+        let inverterStateColour = "";
+        let inverterStateMsg = "";
+        let inverterModel = InverterModel.Sunsynk;
+        let inverterState = stateObj21.state as any;
 
-    if (config.inverter.model !== 'sunsynk' && config.inverter.model !== 'lux' && config.inverter.model !== 'goodwe_gridmode' && config.inverter.model !== 'goodwe' ) {
-      inverterModel = 'sunsynk';
-    } else {
-      inverterModel = config.inverter.model;
-    }
+        let found = false;
 
-    let typeStatusGroups = inverterStatusGroups[inverterModel];
-
-    for (const groupKey of Object.keys(typeStatusGroups)) {
-      const info = typeStatusGroups[groupKey];
-      const { states, color, message } = info;
-      if (states.includes(stateObj21.state)) {
-        inverterStateColour = color;
-        inverterStateMsg = message;
-        found = true;
-        break;
-      }
-    }
-    if (!found) {
-      if (config.entities?.inverter_status_59 === 'none' || !config.entities?.inverter_status_59) {
-        inverterStateColour = 'transparent';
-        inverterStateMsg = '';
-      } else {
-        inverterStateColour = 'transparent';
-        inverterStateMsg = 'Status';
-      }
-    }
+        // Check if the userInputModel is a valid inverter model
+        if (Object.values(InverterModel).includes(config.inverter.model)) {
+            inverterModel = config.inverter.model as InverterModel;
+        }
 
-    //Set Battery Status Message and dot for goodwe
-    let batteryStateColour = "";
-    let batteryStateMsg = "";
-
-    let batStatusGroups = batteryStatusGroups[inverterModel];
-
-    if (inverterModel === 'goodwe_gridmode' || inverterModel === 'goodwe') {
-      for (const groupKey of Object.keys(batStatusGroups)) {
-        const info = batStatusGroups[groupKey];
-        const { states, color, message } = info;
-        if (states.includes(stateObj66.state)) {
-          batteryStateColour = color;
-          batteryStateMsg = message;
-          found = true;
-          break;
+        if (grid_voltage != null && !isNaN(grid_voltage)) {
+            grid_status = grid_voltage > 0 ? 'on' : 'off';
         }
-      }
-      if (!found) {
-        if (config.entities?.battery_status === 'none' || !config.entities?.battery_status) {
-          batteryStateColour = 'transparent';
-          batteryStateMsg = '';
-        } else {
-          batteryStateColour = 'transparent';
-          batteryStateMsg = 'Status';
+        if (battery_current_direction != null && !isNaN(battery_current_direction)) {
+            if (inverterModel == InverterModel.Solis && battery_current_direction === 0) {
+                battery_power = -battery_power;
+            }
```

Comment:
Is battery power always reported as a positive number for the Solis Inverter regardless of whether you are charging or discharging ?  Trying to understand why you need this?

Reply:
Should the logic not be:

if (inverterModel == InverterModel.Solis && battery_current_direction >= 0)

Reply:
Hi, yes always a positive value, the direction has 2 states, 0 and 1. Where 0 is for discharging